GTScript-Parallel-model.md

```diff
@@ -134,17 +180,233 @@ The computation domain of every statement is extended to ensure that any require
+
+There are different uses in which writing off-center can work with clear semantics: write at a location before the iteration loop reaches it (e.g., offset greater than zero in a FORWARD loop), and write at a location after the iteration loop reaches it (e.g., offset less than zero in a FORWARD loop). Both cases are valid. In the first we want to make the values visible in the same computation, in the second case we want to make the values available for subsequent computations. There are cases with conditionals in which the two cases above can be mixed up, but they are deterministic and creates no problems, even though understanding the code by the programmers can be difficult.
+
+### Policy for off-center writes
+
+The policy is the following:
+
+- If iteration policy is unspecified then write off-center in the vertical direction is forbidden and raise a compile error
+- If iteration policy is not PARALLEL the write off-center is allowed.
+
+In all this discussion we assumed the fields where off-center writes were performed had been allocated with the proper sizes and index spaces to accommodate them.
+
+<details>
+### Simple example
+We introduce the behavior and policies in GTScript by looking at one sample code executed the in three different iterations policies.
+
+```python
+with computation(FORWARD):
+    with interval(start, end):
+        a[0,0,-1] = c
+        b[0,0,1] = 2 * a # b will see old values of a
+
+with computation(BACKWARD):
+    with interval(start, end):
+        a[0,0,-1] = d
+        b[0,0,1] = 2 * a # b will see newly updated values form d (apart from the first k level (end))
+
+with computation(...):
+    with interval(start, end):
+        a[0,0,-1] = c
+        b[0,0,1] = 2 * a # b will have unspecified values depending on iteration order
 ```
+
+This code is translated in pseudo-code as:
+
+```python
+for k in range(start, end):
+    a[i:I,j:J,k-1] = c[i:I,j:J,k]
+    b[i:I,j:J,k+1] = 2 * a[i:I,j:J,k] # b contains 2 times the previous values in a before the assignment of c, since a and b are evaluated at the same k-level
+
+for k in range(end, start):
+    a[i:I,j:J,k-1] = d[i:I,j:J,k]
+    b[i:I,j:J,k+1] = 2 * a[i:I,j:J,k] # b is going to contain the 2*d except for the firt iteration point, max2.
+
+    # The third case is ambiguous and will raise a compilation error
+```
+
+### Another code sample with conditionals:
+
+```python
+with computation(FORWARD):  # Forward computation
+    with interval(start, end):
+        if (x < 0):
+            a[0,0,-1] = 1
+        else
+            a[0,0,1] = 2
+
+        b = a
+```
+
+Which translates to
+
+```python
+for k in range(start, end):
+    mask[i:I, j:J] = x < 0
+    parfor ij:
+        if (mask):
```

Comment:
@havogt we had this discussion in the previous PR.
```
@havogt havogt 2 days ago Member
where x<0 in the k-1 slice!!!
I think we should disallow off-center writes in ifs on fields.

@mbianco mbianco 2 days ago Author Member
No, why? x is evaluated in the current slice...
```
I do not see the problem in this example. Could you explain (or we can talk directly if it's easier)

Reply:
Yes, x is evaluated at center but writing to a is off-center which means b will actually even contain either 1 or 2 or the old value of `a` I guess.

Reply:
Let's say we have only a single k column with 
- `mask = [-,-1,1,1,1,-1,-]` (`-` means outside of iteration range)
- `a = [a,a,a,a,a,a,a]`

Then I iterate the indices 1, 2, 3, 4
1. a = [2,a,a,a,a,a,a] (write to k-1)
2. a = [2,a,a,1,a,a,a] (write to k+1)
3. a = [2,a,a,1,1,a,a] (write to k+1)
4. a = [2,a,a,1,1,1,a] (write to k+1)
5. a = [2,a,a,1,2,1,a] (write to k-1)

So result is `b = [2,a,a,1,2,1,a]`

Reply:
I think you missed the point that `b` is only updated at the current level after each iteration, so the `if (x < 0)` branch will never update `b` only `a`, and it will update it following the pattern you expose here

Reply:
Yes makes sense.

Reply:
So the result is `b=[b,a,a,1,1,1,b]`, i.e. b is 1 where x > 0 in the k-1 slice?

Reply:
Did I also mess up the 2 and the 1 in my example?

Reply:
1. a = [1,a,a,a,a,a,a] (write to k-1)
2. a = [1,a,a,2,a,a,a] (write to k+1)
3. a = [1,a,a,2,2,a,a] (write to k+1)
4. a = [1,a,a,2,2,2,a] (write to k+1)
5. a = [1,a,a,2,1,2,a] (write to k-1)

and the result is `b=[b,a,a,2,2,2,b]` i.e. 2 if x> 0 at k-1?

Reply:
If the rules I wrote in the previous comment are correct
```python
a[k] == 1 if a[k+1] < 0
a[k] == 2 if a[k-1] >= 0 and a[i,j,k+1] >= 0
a[k] == a_original[k] in any other point

b[k] == 2 if a[k-1] >=0 else a_original[k]
```
the results should be:
```python
# a
a == 1 at [0, 4, 5?] # 5 would belong here if a[6] < 0
a == 2 at [3, 5?]  # 5 would belong here if a[6] >= 0
a == orig_a at [1,2] 

a = [1, a, a, 2, 1, 1/2?, a]

# b
b == 2 at [3,4,5]
b == orig_a at [1,2]
b == orig_b at [0,6]

b = [b, a, a, 2, 2, 2, b]
```

Reply:
You are in a different example? Here the condition is on x.

Reply:
Oh, yes, you're right, I don't know why, but I was always thinking the condition was on `a`